### ISSUE TITLE
[QUESTION] how to properly sanitize API request payloads

### ISSUE THREAD

**Issue by tezvi** [api-platform/api-platform#1089]
Just a general question what do you guys recommend as a best practice to sanitize user request payloads for XSS and other potential harmful data.

There are several ideas that come on my mind, like:
- Entity fields validation using Symfony Validator and annotations
- Decorating deserializer / denormalizer / decoder and filtering string data from payload requests
- Additional Apache / Nginx handling and filtering

Any other ideas? Does api-platform have it's own way of handling this? It seems to me that data is persisted straight from the request.

**Comment by soyuka** [comment]
Data is validated after the denormalization phase. To avoid type errors you can disable type checking in the normalizers.

**Comment by tezvi** [comment]
I understand. However I was aiming more into a security kind of checking.
For an example I would like to prevent malicious javascript data from even reaching a validation phase and deserialization to an entity.

I've come up with a simple implementation by decorating a default decoder and manually inspecting string values from a raw JSON payload and throwing HTTP 400 if funky business were detected. It's a bit defensive I know so i was wondering about alternatives.

**Comment by teohhanhui** [comment]
1. Always use parameterized (database) queries.
2. Always escape (for the correct context) untrusted data (anything that came from user input) on the client.

**Comment by tezvi** [comment]
Hi, thanks for input.
You are absolutely right, but how do you suggest to accomplish 2nd rule if you are using a default api POST endpoint configuration without controller or any custom listener and only default constraint validators on entity (no fancy custom filtering).
If you send a HTTP POST payload with malicious XSS it will be persisted to database. It means it will go through ReadListener, ValidationListener, WriteListener and end up in database.

**Comment by sclausen** [comment]
If a string containing javascript is stored in the database, it’s not harmful per se. You shouldnt evaluate the javascript on output though.

**Comment by tezvi** [comment]
Of course, that might be the solution. However, by using a default ApiPlatform implementation the XSS content will be returned to client. So you'll have to tweak around that.
Concerning security in general, a proper course of action should be to prevent any malicious data from entering into persisting layer. Better safe than sorry :)

**Comment by soyuka** [comment]
Closing per #2987. Questions are best asked on [GitHub Discussions](https://github.com/api-platform/core/discussions) or [Stack Overflow (api-platform tag)](https://stackoverflow.com/questions/tagged/api-platform.com). The framework now lives at `api-platform/core`.
